File: typescript/src/main/vsdk/toolkit/raycasting/bidirectionalRaytracing/SparList.ts
import { ColorRgb } from "../../common/color/ColorRgb";
import { CircularList } from "../../common/dataStructures/CircularList";
import { CircularListIterator } from "../../common/dataStructures/CircularListIterator";
import { BiPath } from "./BiPath";
import { Spar } from "./Spar";
import { SparConfig } from "./SparConfig";

export class SparList extends CircularList<Spar> {
  public handlePath(config: SparConfig, path: BiPath, fRad: ColorRgb, fBpt: ColorRgb): void {
    const iter = new CircularListIterator<Spar>(this);

    fBpt.clear();
    fRad.clear();

    let spar: Spar | null;
    while ((spar = iter.nextOnSequence()) !== null) {
      const col = spar.handlePath(config, path);

      if (spar === config.leSpar) {
        fBpt.add(col, fBpt);
      }
      else {
        fRad.add(col, fRad);
      }
    }
  }
}

File: java/base/src/main/java/vsdk/toolkit/raycasting/bidirectionalRaytracing/SparList.java
package vsdk.toolkit.raycasting.bidirectionalRaytracing;

import vsdk.toolkit.common.color.ColorRgb;
import vsdk.toolkit.common.color.ColorRgbMutable;
import vsdk.toolkit.common.dataStructures.CircularList;
import vsdk.toolkit.common.dataStructures.CircularListIterator;

public final class SparList extends CircularList<Spar> {
    public void
    handlePath(
        SparConfig config,
        BiPath path,
        ColorRgbMutable fRad,
        ColorRgbMutable fBpt)
    {
        CircularListIterator<Spar> iter = new CircularListIterator<>(this);
        Spar spar;
        ColorRgb col;

        fBpt.clear();
        fRad.clear();

        while ( (spar = iter.nextOnSequence()) != null ) {
            col = spar.handlePath(config, path);

            if ( spar == config.leSpar ) {
                fBpt.add(new ColorRgbMutable(col), fBpt);
            } else {
                fRad.add(new ColorRgbMutable(col), fRad);
            }
        }
    }
}
